WIP: Add zfs vol driver

MatiasVara opened the pull request:
This PR adds a ZFS volume driver. The current work is a work-in-progress. The goal of this PR is only for reviewing. Note that both the architecture and the implementation shall be reviewed. The architecture is described in the design document.

MatiasVara committed:
WIP: Add support for zfs volume driver

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
WIP: Add sr.destroy()

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
WIP: Add vdi.destroy()

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
WIP: Add vdi.stat()

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
WIP: Add sr.ls()

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
WIP: Add vol.snapshot()

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
WIP: Add vol.clone()

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add clone and snapshot in plugin.py

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
VDI_DESTROY fails if volume has children

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add support for revert-from-snap.

The clone method only accepts snapshot. Destroy does not fail
if the volume has children.

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add promote to remove parent when during revert-from-snapshot

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Force creation when blk device is dirty

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Creates a clone during snapshot and removes it when the snapshot is destroyed

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Fix typo in cmd

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add parameters during vdi creation.

This commit adds the following parameters during vdi creation:
- mountpoint
- compression
- devices
- mode

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add promotion of clone during snapshot

This commit adds promoting when a snapshot is taken together with the clone
that is created. This allows to remove the main volume during
revert-from-snapshot. This also allows to revert from any snapshot.

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Do not use parent id during stat

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add support for destroy volumes and snapshots

This commit improves how the driver destroys volumes and snapshots. The
commit creates a tree structure in which snapshots are the parent of volumes
and other snapshots. The root of this tree is the oldest snapshot. The button
of the three is the latest snapshot. Snapshots can be destroy from top to
button (from oldest to lastest) or from button to top (from latest to oldest).
Those snapshots that have volumes cannot be destroyed until all volumes are
destroyed.

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Use COWUtil interface for volume operations

This commit creates zfsutil.py library that contains the class ZFSUtil that
instantiates the interface COWUtil. This interface contains common operations
over volumes. The commit rewrites volume.py and sr.py by using this library.
The commit also extends current COWUtil interface by adding three new
operations:
- promote
- destroy
- clone
These operations are supported by zfs volumes.

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Remove unnecessary constants

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/plugin.py:
```diff
@@ -0,0 +1,50 @@
+    def query(self, dbg):
+        return {
+            "plugin": "zfs-ng",
+            "name": "Zfs Volume plugin",
```
ZFS*

Wescoeur commented on xapi/storage/libs/libcow/zfsutil.py:
```diff
@@ -0,0 +1,57 @@
+        return call(dbg, cmd)
+
+    @staticmethod
+    def destroy(dbg, path):
```
Because it's specific to ZFS, you can add a comment here.

Wescoeur commented on xapi/storage/libs/libcow/zfsutil.py:
```diff
@@ -0,0 +1,57 @@
+        return call(dbg, cmd)
+
+    @staticmethod
+    def promote(dbg, clone_path):
```
Maybe you can add a comment to indicate that this method is specific to ZFS, not used by other drivers.
And also why we need it.

Wescoeur commented on xapi/storage/libs/libcow/zfsutil.py:
```diff
@@ -0,0 +1,57 @@
+
+ZFS_UTIL_BIN = 'zfs'
+
+class ZFSUtil(COWUtil):
```
This method is not implemented:
```python
    @staticmethod
    def getImgFormat(dbg):
        return 'raw'
```
I suppose we must return 'raw' like the `RawUtil` helper.

Wescoeur replied:
Same for `is_empty`, I know it's used in `epc_open`.

MatiasVara replied:
I implemented `getImgFormat()` and `is_empty()`. However, I am not sure what is the meaning of `is_empty()`. I implemented like:
```python
    @staticmethod
    def is_empty(dbg, vol_path):
        raise NotImplementedError()
```

Wescoeur commented on xapi/storage/libs/libcow/zfsutil.py:
```diff
@@ -0,0 +1,57 @@
+    def get_vsize(dbg, vol_path):
+        cmd = [
+            ZFS_UTIL_BIN, 'get',
+            '-o', 'value', '-Hp', 'used,avail',
```
Well I'm not sure why you return two fields here: `used` and `avail`, normally only `used` is useful no? Like the virtual size for VHD images. Also is the information returned using bytes?

MatiasVara replied:
I fixed it by only returning used and by adding a comment that explains that the returned value is in bytes.

Wescoeur commented on xapi/storage/libs/libcow/zfsutil.py:
```diff
@@ -0,0 +1,57 @@
+            path
+        ]
+        return call(dbg, cmd)
+
```
Don't forget to implement these methods:

```python
    @staticmethod
    def online_snapshot(dbg, new_cow_path, parent_cow_path, force_parent_link):
        return ZFSUtil.snapshot(
            dbg, new_cow_path, parent_cow_path, force_parent_link)

    @staticmethod
    def offline_snapshot(
            dbg, new_cow_path, parent_cow_path, force_parent_link):
        return ZFSUtil.snapshot(
            dbg, new_cow_path, parent_cow_path, force_parent_link)
```

I know that they are not used by this driver directly, but they exist, and if the `ZFSUtil` interface should be used elsewhere, it would be nice to have the implementation of these methods as for the other drivers.

MatiasVara replied:
I added these methods.

Wescoeur commented on xapi/storage/libs/libcow/cowutil.py:
```diff
@@ -17,6 +17,34 @@ def is_empty(dbg, vol_path):
+            clone_path: (str) Absolute path to a clone
+        """
+        raise NotImplementedError()
+
```
Not sure if we must add these definitions here. It's very specific to ZFS. To discuss.

MatiasVara replied:
I removed these definitions for the moment.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+
+        # TODO: rewrite this
+        psize = int(out[0]) + int(out[1])
+        fsize = int(out[1])
```
Well, I suppose there a problem in this addition. :sweat_smile:

MatiasVara replied:
I fixed this by just using the used value. `fsize` is equal to psize since for a raw volume the fsize is the whole disk.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+                if is_snapshot:
+                    psize = int(out[0])
+                else:
+                    psize = int(out[0]) + int(out[1])
```
I don't understand this part. :)

MatiasVara replied:
I was wrongly computing the size of volumes. I fixed by just using the used value, i.e., out[0], however for a snapshot this is 0 and only increases when snapshot changes from the original volume.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+            log.debug('error destroying the pool the pool')
+
+    def detach(self, dbg, sr):
+        # Nothing todo.
```
*to do :wink:

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+    def destroy(self, dbg, sr):
+        name = os.path.basename(sr)
+        cmd = [
+            ZPOOL_BIN, 'destroy',
```
Use ZFSUtil?

MatiasVara replied:
ZFSUtil.py:destroy() uses ZFS_BIN instead of ZFS_POOL. That's why I do not use ZFSUtil for ZPOOL_BIN commands.

MatiasVara replied:
I added destroy_pool() in ZFSUtil.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+        try:
+            call(dbg, cmd)
+        except:
+            log.debug('error destroying the pool the pool')
```
You can catch the error and fix the message I suppose. ^^"

MatiasVara replied:
Is it not already catched? I'll fix the msg.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+            raise
+        devs = configuration['devices'].split(',')
+
+        mountpoint = '/' + name
+        if 'mountpoint' in configuration:
+            mountpoint = configuration['mountpoint']
```
Or:

```python
configuration.get('mountpoint', '/' + name)
```

MatiasVara replied:
If line 48 is True, the mountpoint is got from configuration otherwise we use the value sets at line 47. Is it OK like this or an else would make it clear?

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+
+        cmd = [
+            ZPOOL_BIN, 'create', '-f',
+            name, '-m', mountpoint
```
You can use a method defined in ZFSUtil. :)

MatiasVara replied:
ZFSUtil.py:create() uses ZFS_BIN instead of ZPOOL_BIN. That's why I do not use ZFSUtil for ZPOOL_BIN commands.

MatiasVara replied:
I implemented a method named create_pool() in ZFSUtil to do this.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+        compression = False
+        if 'compression' in configuration:
+            if configuration['compression'] in [
+                    'true', 't', 'on', '1', 'yes']:
```
I saw this code in other drivers, maybe we can add a helper for that?

benjamreis replied:
```py
compression = configuration.get('compression') in ['true', 't', 'on', '1', 'yes']
```
?

MatiasVara replied:
I added a helper at util for those cases in which the driver has to check is a variable is on.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+            if configuration['mode'] not in [
+                    'N', 'M', 'R']:
+                log.error('mode can only be N(default), M(mirror) or R(raidz)')
+                raise
```
You can probably use and exception from `xapi.storage.api.v5.volume.XXX`.

MatiasVara replied:
Which exception should I use in this case?

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,226 @@
+
+        if compression:
+            cmd = [
+                ZFS_BIN, 'set', 'compression=on', name
```
Same here, a method defined in ZFSUtil can be useful.

MatiasVara replied:
I added the setcompression() method to ZFSUtil.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,252 @@
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    vdi = db.get_vdi_by_id(key)
+                    if vdi.volume.snap == 0:
```
Or:
```python
if not vid.volume.snap:
```

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,252 @@
+                            child_vol = None
+                            # select any snapshot children to promote
+                            for child in children:
+                                if child.snap == 1:
```
Or:
```python
if child.snap:
```

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,252 @@
+                            ZFSUtil.promote(dbg, path_clone)
+                            db.update_volume_parent(child_vol.id, vdi.volume.parent_id)
+                            # snapshot path changes after promotion
+                            path = os.path.basename(sr) + '/'+ str(child_vol.id) + '@' + str(vdi.volume.id)
```
Space after '/', also maybe you can add an helper for this concatenation.

MatiasVara replied:
I added a helper to build the path to a snap.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,252 @@
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id) + '@' + str(snap_volume.id)
+                    ZFSUtil.clone(dbg, path, os.path.basename(sr) + '/' + str(snap_volume.id))
+                    ZFSUtil.promote(dbg, os.path.basename(sr) + '/' + str(snap_volume.id))
+        psize = 0
```
Why?

MatiasVara replied:
I added a comment to explain why.

MatiasVara replied:
I fixed this by using the value from get_vsize which is the used value. However, in zfs this is always 0 for a snapshot since new snapshots not consume space.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,252 @@
+                    clone_path = os.path.basename(sr) + '/'+ result_volume_id
+                    ZFSUtil.clone(dbg, snap_path, clone_path)
+
+        psize = 0
```
Always 0?

MatiasVara replied:
I fixed this by using the value from get_vsize. I do not remember if clone volumes consumes space.

Wescoeur commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,252 @@
+                    db.update_volume_parent(vdi.volume.id, snap_volume.id)
+
+                    result_volume_id = str(snap_volume.id)
+                    ZFSUtil.snapshot(dbg, str(snap_volume.id),
+                                 os.path.basename(sr) + '/'+ str(vdi.volume.id), False)
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id) + '@' + str(snap_volume.id)
+                    ZFSUtil.clone(dbg, path, os.path.basename(sr) + '/' + str(snap_volume.id))
+                    ZFSUtil.promote(dbg, os.path.basename(sr) + '/' + str(snap_volume.id))
```
An helper to compute paths seems good + a temporary variable. :)

MatiasVara replied:
I added a helper function and a variable.

MatiasVara committed:
Address comments in review

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Add support for destroying snapshot with a single volume child

This commit deals with the case in which a snapshot with a single volume child
is destroyed. To do so, the volume is first promoted, then the clone is
destroyed and finally the snapshot can be destroyed.

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Store mounpoint when creating a sr

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Use name instead of mountpoint in Destroy

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

MatiasVara committed:
Use first snapshot child during destroy

Signed-off-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>

ydirson committed:
Fix error cases

ydirson committed:
Use "sr-{uuid}" instead of label to name the zpool 1/2

And mount it under /var/run/sr-mount/ not / for consistency with other
SRs.

* characters valid in XAPI SR name is not the same as those valid for
  a ZFS pool name, so use UUID instead (would not be able to use under
  xcp-ng-tests otherwise)
* ZFS requires the pool name to start with a letter, so use "sr-" as
  prefix

Just enough for sr-create/destroy, but breaks all volume manipulation.

ydirson committed:
Use "sr-{uuid}" instead of label to name the zpool 2/2

This part deals with the volume manipulation.  Since the volume plugin
does not use a "SR metadata" struct we have to introduce a separate
helper function here - and since we mostly care about it when
accessing a volume, simply handle creation of volume name in one
place.

This is unfortunately needed in other loosely-related code units, so
hardcoding goes on...

ydirson committed:
util.call*: fix error reporting

Trying to interpret the exit status of arbitrary called processes as an
errno value, and disregard the stderr content we extract for a reason, had
little chance to be useful.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>
(cherry picked from commit 2b6fb6c26209e24050f3ded212bc1379ea6ee3d4)

ydirson commented:
Included-in/superceded-by #17, now merged!
